### DISCUSSION TITLE
How to create a Custom Active Model from a Model with an Option<> for a PK?

### DISCUSSION THREAD

**Discussion by benze** [SeaQL/sea-orm#1512] [Q&A]
I noticed that in the ActiveModel.save() function, that if the ActiveModel PK is `NotSet`, the method performs an insert.  If the PK is set, the method performs an update.

This is great when working with the ActiveModel directly.  However, I wanted to expand that to a CustomActiveModel derivation so that my incoming APIModel could have the PK as an Option.

For example:
```
#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq)]
#[sea_orm(table_name = "emails")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i32,
    pub user_id: Option<i32>,
    pub email_id: Option<i32>,
    pub mailbox: Option<String>,
    pub downloaded_at: DateTimeWithTimeZone,
    #[sea_orm(column_type = "Text", nullable)]
    pub raw_contents: Option<String>,
}


#[derive(DeriveIntoActiveModel)]
pub struct ApiModel {
    pub id: Option<i32>,
    pub user_id: Option<i32>,
    pub downloaded_at: DateTimeWithTimeZone,
    pub raw_contents: Option<String>,
}
```

The problem here is that the compiler complains that there is no From<Option<i32>> to ActiveValue<i32> trait.
```
6 | #[derive(DeriveIntoActiveModel)]
  |          ^^^^^^^^^^^^^^^^^^^^^ the trait `std::convert::From<ActiveValue<Option<i32>>>` is not implemented for `ActiveValue<i32>`
  |
  = help: the trait `std::convert::From<ActiveValue<V>>` is implemented for `ActiveValue<Option<V>>`
  = note: required for `ActiveValue<Option<i32>>` to implement `Into<ActiveValue<i32>>`
  = note: this error originates in the derive macro `DeriveIntoActiveModel` (in Nightly builds, run with -Z macro-backtrace for more info)
```

My goal is to be able to do something simple like:
```
async fn save( model: ApiModel, db : &DatabaseConnection ){
   model.into_active_mode().save(&db).await;
}
```

Using this type of ApiModel allows me to either set the `id` (PK) or not.  If the id is set, the post method ends up doing an update.  If it is not, it ends up doing an insert.

The alternative that I see it to use 2 separate models - one for a save and one for an update, but that doesn't really appeal to me.   Similarly, I'm not fond of using the ActiveModel directly in my business logic either.  So the business logic layer would need 2 representations of the same object which seems cumbersome.

I could manually create the ActiveModel by coping fields individually from the ApiModel, but that defeats the purpose of the `into_active_model()` method. And once the save() is complete, how would I go back to mapping it back to an ApiModel?  Again would seem to be a manual process of mapping field by field.

What kind of approach can I use for this?

**Comment by tyt2y3** [comment]
I would suggest you to use two entities then - one trimmed to satisfy the business logic.
In your case, it might even make sense to have two entity crates, one external facing, one for administration.

**Comment by Winor** [comment]
What would be the best approach in case I have default values set in my database and I don't want to force the user to always have to provide all the keys while inserting data? 

For example 

item.rs:
```rust
//! `SeaORM` Entity. Generated by sea-orm-codegen 0.11.0
use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "items")]
pub struct Model {
    #[sea_orm(primary_key)]
    #[serde(skip_deserializing)]
    pub id: i32,
    pub name: String,
    pub quantity: i32, // defaults to 1
    pub unit: String, // defaults to "piece"
    #[sea_orm(column_type = "Text", nullable)]
    pub description: Option<String>,
    pub instock: bool, // defaults to true
}
```
It would've been cool if I could do something like this:
user_input.rs
```rust
use serde::{Deserialize, Serialize};
use sea_orm::entity::prelude::*;
use crate::items::ActiveModel;


#[derive(Serialize, Deserialize, DeriveIntoActiveModel)]
pub struct NewItem {
    pub name: String,
    pub quantity: Option<i32>,
    pub unit: Option<String>,
    pub description: Option<String>,
    pub instock: Option<bool>,
}
```
But unfortunately, it doesn't seem to work like that (for the same reason)
What would be the best approach in this case? Sorry if I'm missing something ...

**Reply by Winor** [reply]
I solved this by making the cols that have default value in the database nullable :)

item.rs:
```rust
#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "items")]
pub struct Model {
    #[sea_orm(primary_key)]
    #[serde(skip_deserializing)]
    pub id: i32,
    pub name: String,
    pub quantity: Option<i32>,
    pub unit: Option<String>,
    #[sea_orm(column_type = "Text", nullable)]
    pub description: Option<String>,
    pub instock: Option<bool>,
}
```

**Comment by Winor** [comment]
While the workaround described in my post above works for the case of needing to `insert` from user input, it is not ideal for `update` because again, I don't want to force the user to always have to provide all the keys. I noticed that `DeriveIntoActiveModel` won't accept setting `Option<T>` to a required field, so another workaround I found is making all fields but the primary key optional and later decide if a field is required or not using a custom `ActiveModel` because  `DeriveIntoActiveModel` will allow me to convert  `Option<T>` to 'T' but not the other way around.
My question, is there any better way to achieve this?
